src/blockdata/block.rs

```diff
@@ -58,6 +58,61 @@ pub struct Block {
     pub txdata: Vec<Transaction>
 }
 
+impl Block {
+    /// check if merkle root of header matches merkle root of the transaction list
+    pub fn check_merkle_root (&self) -> bool {
+        self.header.merkle_root == self.merkle_root()
+    }
+
+    /// check if witness commitment in coinbase is matching the transaction list
+    pub fn check_witness_commitment(&self) -> bool {
+        let witness_root = self.witness_root();
+        // witness commitment is optional if there are no transactions using SegWit in the block
+        if self.merkle_root() == witness_root {
```

Comment:
Hmm? The witness_root() skips the coinbase tx, so this can never match, no? Also, would be much faster to actually check for the presence of witnesses in transactions.

Reply:
you are right. fixing.